Library/src/main/java/com/shopify/testify/internal/helpers/OrientationHelper.kt

```diff
@@ -0,0 +1,122 @@
+import androidx.test.rule.ActivityTestRule
+import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry
+import androidx.test.runner.lifecycle.Stage
+import com.shopify.testify.internal.exception.UnexpectedOrientationException
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+internal class OrientationHelper<T : Activity>(
+    private val activityClass: Class<T>
+) {
+    var requestedOrientation: Int? = null
+    private lateinit var rule: ActivityTestRule<T>
+    private lateinit var lifecycleLatch: CountDownLatch
+
+    fun afterActivityLaunched(rule: ActivityTestRule<T>) {
+        this.rule = rule
+        this.requestedOrientation?.let {
+            if (!activity.isRequestedOrientation(it)) {
+                activity.changeOrientation(it)
+            }
+        }
+    }
+
+    /**
+     * Validate that the activity's orientation matches what was requested
+     */
+    fun assertOrientation() {
+        requestedOrientation?.let {
+            if (!activity.isRequestedOrientation(it)) {
+                throw UnexpectedOrientationException("Device did not rotate.")
+            }
+        }
+    }
+
+    fun afterTestFinished() {
+        requestedOrientation = null
+    }
+
+    private val activity: T
+        get() {
+            return rule.activity
+        }
+
+    private val Activity.isLandscape: Boolean
+        get() {
+            val size = Point(-1, -1)
+            this.windowManager?.defaultDisplay?.getRealSize(size)
+            return size.y < size.x
+        }
+
+    /**
+     * Check if the activity's current orientation matches what was requested
+     */
+    private fun Activity.isRequestedOrientation(requestedOrientation: Int): Boolean {
+        return (requestedOrientation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE && this.isLandscape) ||
+                (requestedOrientation != ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE && !this.isLandscape)
+    }
+
+    /**
+     * Lifecycle callback. Wait for the activity under test to completely resume after configuration change.
+     */
+    private fun lifecycleCallback(activity: Activity, stage: Stage) {
+        if (activity::class.java == activityClass) {
+            if (stage == Stage.RESUMED) {
+                lifecycleLatch.countDown()
+            }
+        }
+    }
+
+    private fun Activity.changeOrientation(requestedOrientation: Int) {
+        ActivityLifecycleMonitorRegistry.getInstance().addLifecycleCallback(::lifecycleCallback)
+
+        Espresso.onIdle()
+
+        val rotationLatch = CountDownLatch(1)
+        lifecycleLatch = CountDownLatch(1)
+
+        this.runOnUiThread {
+            this.requestedOrientation = requestedOrientation
+            rotationLatch.countDown()
+        }
+
+        // Wait for the rotation request to be made
+        if (!rotationLatch.await(1, TimeUnit.SECONDS)) {
+            throw UnexpectedOrientationException("Failed to apply requested rotation.")
+        }
```

Comment:
failing to understand this. we update an int var on the ui thread and then unlock the latch. what actually causes the the config change here

Reply:
@fredells `requestedOrientation` is the kotlin property syntax for `setRequestedOrientation`
It's a method, not a property
It calls this https://developer.android.com/reference/android/app/Activity#setRequestedOrientation(int)